sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java

```diff
@@ -239,8 +246,53 @@ private TestPipeline(
   public Statement apply(final Statement statement, final Description description) {
     return new Statement() {
 
+      private final Predicate<Annotation> hasCategoryAnnotation = new Predicate<Annotation>() {
+
+        @Override
+        public boolean apply(final Annotation annotation) {
+          return annotation.annotationType() != null
+              && annotation.annotationType().equals(Category.class);
+        }
+      };
+
+      private final Predicate<Annotation> categoryOfNeedsRunner = new Predicate<Annotation>() {
+
+        @Override
+        public boolean apply(final Annotation category) {
+          return
+              FluentIterable
+                  .from(Arrays.asList(((Category) category).value()))
+                  .anyMatch(new Predicate<Class<?>>() {
+
+                    @Override
+                    public boolean apply(final Class<?> aClass) {
+                      return NeedsRunner.class.isAssignableFrom(aClass);
+                    }
+                  });
+        }
+      };
+
+      private void setDeducedEnforcementLevel() {
+        if (enforcement instanceof PipelineAbandonedNodeEnforcement) {
```

Comment:
That's a really good point. 

First, if a user changes the enforcement level inside the test code, it will take effect since we call ` statement.evaluate()` (i.e. the user code) **after** `setDeducedEnforcementLevel` is called.
```java
@Override
      public void evaluate() throws Throwable {
        setDeducedEnforcementLevel(); // set the enforcement level we think should be set
        statement.evaluate(); //user code is run here, so any changes will overwrite our pervious setting for the enforcement level
        enforcement.afterTestCompletion();
      }
```

If the user changes the enforcement level at the rule level, i.e. outside of the test, e.g.  `@Rule TestPipeline = TestPipeline.create().enableAbandonedNodeEnforcement(...)` it is handled inside `setDeducedEnforcementLevel()`
```java
if (enforcement instanceof PipelineAbandonedNodeEnforcement) { ... }
 // if enforcement instanceof PipelineRunEnforcement at this point, it must have been
// set so by the user, so don't second guess it.
```

Basically this should cover both cases of users change enforcement level inside and outside their tests.

Do you believe we have other use-cases that need handling?